aiohttp/web.py

```diff
@@ -272,6 +274,20 @@ def _make_request(self, message, payload, protocol, writer, task,
             self._loop,
             client_max_size=self._client_max_size)
 
+    def _prepare_middleware(self):
+        for m in reversed(self._middlewares):
+            sig = inspect.signature(m)
```

Comment:
I don't like signature inspecting. The check is too weak.
In my mind we should mark a new style middleware very explicitly.
Let's wrap it by `@newstylemiddleware` decorator.
The decorator will add `__middleware_version__` attr to decorated object with value `1`.

@samuelcolvin what do you think?

Reply:
Seems ugly, but I agree the signature approach is brittle.

I'll add it unless anyone can think of a third, less cleaner approach?

Reply:
instead of the decorator we could use a type annotation for new style middleware? So if you want to use new style middleware you have to use 

```python
def my_middleware(request, app) -> Request:
    ...
```

After all, this is what type annotations are for.

Reply:
We could then add a depeciation warning in future if middleware doesn't include a type annotation declaring its return type.

Reply:
Yep, type annotations is even better than decorator!

Reply:
Sure, your snippet is incorrect but I've gotten your idea

Reply:
Hmm, after careful thinking I have objections for type annotations usage.
It's intended for static analysis, processing types information in runtime doesn't sound like good idea.
Moreover supporting https://www.python.org/dev/peps/pep-0563/ is complicated, extracting types information from `.pyi` files and other sources opens a can of worms.

For sake of performance we need checking for middleware version on application freeze stage and storing a flag somewhere anyway.

Reply:
How about to not change signature so dramatically? If function takes two arguments - ok, that's old style. If three - new one. Application as a standalone parameter is very handy in middlewares in anyway.

Reply:
The intentions for type hinting are deliberately vague, I would say the zen was "use it for whatever you want". For example I use it a lot in [pydantic](https://github.com/samuelcolvin/pydantic) and it works well.

O don't think there's anything complicated with extracting information etc, just:

```python
 class Foo:
     pass
 
 async def foo(a, b) -> Foo:
     pass
 
 foo.__annotations__
 {'return': __main__.Foo}
```

@kxepal extra argument out would be another solution, but I think it's kind of unclear and prone to confuse people.

Reply:
Check for return value type only is too weak: both new and old style middlewares return a response:
```def old_style(app: Application, handler: Awaitable) -> StreamResponse```
and
```def new_style(request: Request, handler: Awaitable) -> StreamResponse```
check for arguments count is also not strict:
```def new_style(self, request: Request, handler: Awaitable) -> StreamResponse```

Reply:
@asvetlov the self is not a problem if you inspect class instance method rather than class method itself.

Reply:
Ok, how about different way: currently middleware definition is basically hand-written decorator. What if we introduce own middleware decorator as @asvetlov proposed early, but no need to do anything private attributes. It's ours decorator which demands some protocol and expects that decorated function will accept two specific arguments. Like what `signal` module does.

Reply:
Yep, it's true.
Another important point is supporting default values:
```def new_style(app, request, handler, param=None)```.
We cannot protect it, users will make very wild signatures.

Reply:
(I replied via email and my message came in the wrong place, so I'ved deleted it and am replying here).

@asvetlov I think you have the signature confused.

Old style signature:
```python
def old_style(app: Application, handler: Awaitable) -> Awaitable
    ...
```

New style signature:
```python
def old_style(app: Application, handler: Awaitable) -> StreamResponse
    ...
```

They're completely different. The return type alone is enough to distinguish the two styles of middleware.

Reply:
```
>>> from aiohttp import StreamResponse
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'StreamResponse'
```

From where I should import it? Why I should import it if I don't use it? I think annotations shouldn't be made mandatory. Python itself doesn't requires them to be mandatory to achieve any profits, so we cannot demand them from users.

Reply:
why not?

It's there as a language feature, we can use it if we want.

We can implement the decorator solution too in case some people can't (eg. 3.4) or don't want to use type annotations.

Reply:
@samuelcolvin yes, new and old styles are different by return type, now I see it.

The language feature's usage is static type checking only. Technically you could use it for different purposes but this is discouraged.
Strings as type annotations are supported now by `typing` and will become mandatory in next Pythons. Stub files make situation even more complex.
Let's use decorator approach.
@kxepal could you explain your idea?

Reply:
Genuine question: "but this is discouraged", can you point me to where that's stated? I wasn't aware of it.

I really don't see that stub files have have relevance here, AFAIK they're particular to mypy. I was just suggesting using `function.__annotations__`, same as another application might use `function.__name__`. Stub files and strings as type annotations have no bearing on this.

However, I see a real point to not using type annotations: developers assume they're either advisory or for static type analysis; therefore it doesn't make much sense to use them for business logic in one place in the code.

I'll implement the decorator solution now.

Reply:
@samuelcolvin I don't remember exact blogpost or article but Guido told many times about his vision for annotations: previously it was not utilized by python stdlib etc., people used to apply them on they own.
But now annotations are for type checking only. Other approaches are still supported to not break existing code (and will be supported at least for decade) though.

Reply:
@asvetlov 

```
def old_style_middleware(app, handler):
  async def middleware_handler(request):
    ...
  return middleware_handler

@middleware
def new_style_middleware(request, handler):
  ...

```

where `middleware` is  something like
```
def middleware(func):
  def old_style_middleware(app, handler):
    async def middleware_handler(request):
        return await func(request, handler)
    return middleware_handler
  return old_style_middleware
```

Or whatever it should be to create new style arguments. But, since aiohttp implements `middleware` decorator, it may dictate what it should decorate and what to pass for decorated function.

Thought, I still not quite follow what's wrong with current approach. Yes, it's a bit verbose, but quite simple and explicit.